lib/CL/devices/pthread/pthread.c

```diff
@@ -117,6 +113,10 @@ struct data {
   mem_regions_management* mem_regions;
 #endif
 
+  /* Lock for command list related operations */
+  pthread_mutex_t cq_lock __attribute__ ((aligned (CACHELINE_SIZE)));
```

Comment:
"Flushs" (flushes) was an important word you missed in your quote. Surely if you need to share a variable you go to lower levels of the cache, but not necessarily to the main (external) memory. 

What you are describing is the general problem of "false sharing" which has nothing to do with volatile as such, although volatile here might cause unneccessarily not keeping the i in register if you guard its updates with a critical section anyways, and makes the false sharing problem worse. 

If you don't need to share int i or you guard it with a mutex (the lock) then it should not be marked volatile, but allow compiler to utilize registers for temporary accesses _inside the critical section_. The false sharing might itself be a problem here which can be cured by padding, but dropping volatile means you don't need to share i, you use atomic updates, or guard it with the mutex (which ensures the variables are flushed from registers after the section), or you have a race.